src/node_file.cc

```diff
@@ -2106,6 +2106,175 @@ static void OpenFileHandle(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+// TODO(@anonrig): Implement v8 fast APi calls for `cpSync`.
+static void CpSync(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  // src, dest, preserveTimestamps, errorOnExist, force, recursive
+  CHECK_EQ(args.Length(), 6);
+  BufferValue src(env->isolate(), args[0]);
+  CHECK_NOT_NULL(*src);
+  ToNamespacedPath(env, &src);
+  THROW_IF_INSUFFICIENT_PERMISSIONS(
+      env, permission::PermissionScope::kFileSystemRead, src.ToStringView());
+
+  BufferValue dest(env->isolate(), args[1]);
+  CHECK_NOT_NULL(*dest);
+  ToNamespacedPath(env, &dest);
+  THROW_IF_INSUFFICIENT_PERMISSIONS(
+      env, permission::PermissionScope::kFileSystemWrite, dest.ToStringView());
+
+  bool preserveTimestamps = args[2]->IsTrue();
+  bool errorOnExist = args[3]->IsTrue();
+  bool force = args[4]->IsTrue();
+  bool recursive = args[5]->IsTrue();
+
+  using copy_options = std::filesystem::copy_options;
+  using file_type = std::filesystem::file_type;
+
+  std::error_code error_code{};
+  copy_options options = copy_options::copy_symlinks;
+
+  // When true timestamps from src will be preserved.
+  if (preserveTimestamps) options |= copy_options::create_hard_links;
```

Comment:
is this correct? 

afaict, this will create hard links in the destination directory to the original files. so, while it will appear upon inspecting the filesystem that the files have been copied and the timestamps have been preserved, the files will not have actually been copied, just linked as directory entries in the filesystem to the same file. 

this means if you change any of the files in the source directory subsequently, the "files" in the destination will also change. i haven't looked at existing code in detail but am pretty sure this is not how it works with this flag. 

i am also guessing without checking that this will only work if source and destination are on same filesystem? :thinking: 

if i am indeed correct, i guess simplest thing to do for now would be to use the slow path if "preserveTimestamps" is set? 

from a quick google, i'm not sure there is an efficient way to copy and preserve timestamps without having to touch each file in a syscall and it just doesn't seem that there is an option rn in std::filesystem to emulate the current behaviour. happy to be wrong though - not a C++ expert.

https://en.cppreference.com/w/cpp/filesystem/copy_options
https://en.wikipedia.org/wiki/Hard_link

Reply:
i put some benches using the test/fixtures/copy directory from node.js and the same options as above here if anyone wants to try them out - should be easy to get them working on linux or macos.
https://github.com/just-js/lo-bench/blob/main/fs/README.md

some things to note:

on linux core i5, using /dev/shm to reduce filesystem/disk overhead

- bun is roughly 4x the performance of current node.js 22.3
- a "zero-overhead" implementation on v8, using lo runtime and v8 fastcalls to call std::filesystem::copy is ~2.5x the performance of current node - can assume this would be the theoretical max of node.js with this change merged i think.
- bun is still roughly 1.5x the optimal v8 implementation using filesystem::copy

on macos m1 mini, using a ram disk

- bun is rougly 2.5x performance of current node.js 22.3
- the "optimal" v8 implementation using fileystem::copy is ~1.3x node.js 22.3
- bun is still roughly 2x better than "optimal" v8 using filesystem::copy

also, for the same workload:

- node does ~90k syscalls
- bun does ~55k
- std::filesystem::copy does ~65k

so, without testing with other directory structures and sizes/depths but assuming we see similar results, we could assert the following:

- this change will still be significantly slower than bun
- there is still no good solution for getting individual file errors back, which seems necessary
- it can't be used when we need to preserve timestamps (see above)
- imo, the existing implementation could probably be changed to not make as many unnecessary syscalls, remain in JS and be within a couple of percentage points of throughput of the bun implementation. this might be a better long term approach?

another couple of things to note
- if we use force: false then the difference between node.js and bun is much smaller
- the bigger/deeper the directory is, the smaller the difference will be between node and bun as the throughput is syscall heavy.
- the difference will probably be less again on an actual filesystem. i just used ramdisk in the benches to eliminate that overhead and get a clearer picture of runtime overhead.

let me know if i got anything wrong - i ran through this pretty quickly, but it should be easy to reproduce those results using the link above. :pray:

Reply:
Sounds good.

For now, I'll split this pull-request into multiple changes, and try to optimize the existing implementation before moving it into full C++.

https://github.com/nodejs/node/pull/53612

Reply:
Here's a different approach: https://github.com/nodejs/node/pull/53614